### PR TITLE
Extract collector for metrics extensibility

### PR DESCRIPTION
**Pull request by i7an**
Split `Publisher` class into two: `Publisher` and `Collector`.

`Collector` is responsible only for collecting metrics therefore it is easy to extend with custom metrics:

```ruby
class CustomCollector < Sidekiq::CloudWatchMetrics::Collector
  def collect
    metrics = super

    metrics << {
      metric_name: 'MaxQueueLatency',
      timestamp: Time.current,
      value: Sidekiq::Queue.all.map(&:latency).max,
      unit: 'Seconds'
    }

    metrics
  end
end

Sidekiq::CloudWatchMetrics.enable!(collector: CustomCollector.new)
```

### CODE REVIEW

File: lib/sidekiq/cloud_watch_metrics/publisher.rb [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                   collector: nil,
+                   **collector_kwargs)
+      # Sidekiq 6.5+ requires @config, which defaults to the top-level
+      # `Sidekiq` module, but can be overridden when running multiple Sidekiqs.
+      @config = config
+      @client = client
+      @namespace = namespace
+
+      if collector_kwargs.size > 0
+        warn <<~TEXT
```
Comment (i7an) [inline lib/sidekiq/cloud_watch_metrics/publisher.rb:34]: I suggest to remove the support for `collector_kwargs` in the future and complete the deprecation in 2.7.

**Comment by i7an** [comment]
@sj26 ping ping

**Comment by i7an** [comment]
@sj26 ping ping ping

**Comment by i7an** [comment]
No response. Closing.

**Comment by gstokkink** [comment]
@sj26 I think this functionality is quite useful for extending the default metrics, what do you think? I've now forked your gem to add this, but I think it could be useful to others, as well.

**Comment by i7an** [comment]
@sj26 If there is any interest in this change I'll revive the PR. Please ping me.

**Comment by gstokkink** [comment]
@sj26 reminder 😃

**Comment by sj26** [comment]
This is a nice idea, and I'm very happy it's working for you, but this gem is intended to be a plug-and-play component not a framework for implementing your own collector :pray:

**Comment by i7an** [comment]
@sj26 This patch makes it easier to extend metrics if you need to, without changing the intended interface of the gem. Such refactoring wont affect the plug-and-play aspect of the gem.

Generally speaking ruby is super flexible so it still possible to extend collected metrics with monkey patching if there is no other choice.

**Comment by gstokkink** [comment]
@sj26 IMO i7an has a good point, I think separating the collector and the publisher is a good idea regardless, in terms of software architecture. Makes it much easier for other folks/forks to add extra metrics and the likes, without having to perform painful rebases each time you create a new release. But up to you of course!

**Comment by gstokkink** [comment]
@sj26 no way to change your mind? 😉

**Pull request closed without merging**